File: shared/src/main/java/ru/privatenull/pnauth/kernel/service/ServiceKey.java
package ru.privatenull.pnauth.kernel.service;
public record ServiceKey<T>(String namespace, String name, Class<T> type) {
    public ServiceKey {
        if (namespace == null || !namespace.matches("[a-z0-9_.-]+")) throw new IllegalArgumentException("invalid namespace");
        if (name == null || !name.matches("[a-z0-9_.-]+")) throw new IllegalArgumentException("invalid service name");
        java.util.Objects.requireNonNull(type, "type");
    }
    public static <T> ServiceKey<T> of(String namespace, String name, Class<T> type) {
        return new ServiceKey<>(namespace, name, type);
    }
    public String id() { return namespace + ":" + name; }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/kernel/service/ServiceKey.kt
package ru.privatenull.pnauth.kernel.service

import java.util.Objects

data class ServiceKey<T>(val namespace: String, val name: String, val type: Class<T>) {
    init {
        require(namespace.matches(Regex("[a-z0-9_.-]+"))) { "invalid namespace" }
        require(name.matches(Regex("[a-z0-9_.-]+"))) { "invalid service name" }
        Objects.requireNonNull(type, "type")
    }

    fun id(): String = "$namespace:$name"

    companion object {
        @JvmStatic
        fun <T> of(namespace: String, name: String, type: Class<T>): ServiceKey<T> {
            return ServiceKey(namespace, name, type)
        }
    }
}
